src/v/cluster/archival_metadata_stm.cc

```diff
@@ -684,159 +692,75 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
-            // `do_replicate_commands` method.
-            if (res || _last_replicate->term < _raft->term()) {
-                _last_replicate = std::nullopt;
-            }
-
-            if (!res) {
-                vlog(
-                  _logger.warn,
-                  "Replication failed for archival STM command: {}",
-                  res.error());
-
-                co_return false;
-            }
-        } catch (const ss::timed_out_error&) {
-            vlog(
-              _logger.warn,
-              "Replication wait for archival STM timed out (timeout = {})",
-              timeout);
-            co_return false;
-        } catch (...) {
-            vlog(
-              _logger.error,
-              "Replication failed for archival STM command: {}",
-              std::current_exception());
-            co_return false;
-        }
-    }
-
-    if (!_raft->is_leader()) {
-        co_return false;
-    }
-
-    const auto last_applied_term = _insync_term;
-    const auto last_applied_offset = _manifest->get_insync_offset();
-    const auto current_term = _raft->term();
-    const auto committed_offset = _raft->committed_offset();
-
-    if (
-      last_applied_term == current_term
-      && last_applied_offset < committed_offset) {
-        // Case 1: we have already synced (once or more) during the current
-        // term, but need to sync again (e.g. re-starting the archiver to
-        // apply topic configs)
-        vlog(
-          _logger.debug,
-          "Syncing archival STM from last applied offset {} to committed "
-          "offest {} in term {}",
-          last_applied_offset,
-          committed_offset,
-          current_term);
-
-        const bool synced = co_await wait_no_throw(
-          committed_offset, model::timeout_clock::now() + timeout);
+    return sync(timeout, nullptr);
+}
 
-        if (!synced) {
-            vlog(
-              _logger.warn,
-              "Failed to sync archival STM from offset {} to offest {} in "
-              "term {}",
-              last_applied_offset,
-              committed_offset,
-              current_term);
-        }
+ss::future<bool> archival_metadata_stm::sync(
+  model::timeout_clock::duration timeout, ss::abort_source* as) {
+    return _lock
+      .with(timeout, [this, timeout, as] { return do_sync(timeout, as); })
+      .handle_exception_type(
+        [](const ss::semaphore_timed_out&) { return false; });
+}
 
-        co_return synced;
-    } else {
-        // Case 2: we have not synced since the last term (e.g. on
-        // leadership changes)
-        vlog(
-          _logger.debug,
-          "Syncing archival STM to latest term {} from term {}",
-          current_term,
-          last_applied_term);
-        const bool synced = co_await persisted_stm::sync(timeout);
+ss::future<bool> archival_metadata_stm::do_sync(
+  model::timeout_clock::duration timeout, ss::abort_source* as) {
+    co_await raft::persisted_stm<>::sync(timeout);
```

Comment:
Shouldn't we handle the result here? Persisted stm sync will return false in couple of cases like not a leader or timeout

Reply:
fixed